File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    #답으로 실패율이 나와야하는 것.
    #실패율이 높은 스테이지부터 내림차순으로 스테이지의 번호가 담겨있는 배열을 return 하도록 solution 함수를 완성하라
    #동점처리 고려해야함
        
        
        ###########
        #스테이지별 도전자 수를 구함
    challenger = [0] * (N+2) #왜냐면 스테이지는 1부터니까.. 0번 인덱스 버리고 1번부터 쓴대?? challenger의 1에는 1번스테이지에 몇명이냐, 2에는 2번스테이지에 머무는 수... 챌린지의 n+1에는 n번 스테이지를 클리어한 사람 수가 들어가게! (0번인덱스 버리고 n+1까지 필요)
        
    for stage in stages:
        challenger[stage]+=1
        
        #스테이지별 실패 도전자 수 계산
    fails={}
    total = len(stages)
    for i in range(1, N+1):
        if challenger[i]==0: #도전한 사람 없음, 실패율 = 0
            fails[i] = 0 #fails는 딕셔너리, i는 key고 0이 value
                
        else:
            fails[i] = challenger[i]/total #실패율
            total -= challenger[i] #다음 스테이지 실패율을 구하기 위해 현재 스테이지 인원수를 뺌
        
    #딕셔너리 정렬하는 법 알고있어야함 필수
    res = sorted(fails, key=lambda x:fails[x], reverse=True)
    
    
    return res

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        // int[] answer = {};
        
        // count - 해당 스테이지에 사람들이 몇 명이나 멈춰있는지, 1부터 N+1까지의 스테이지를 담음
        int[] count = new int [N + 2];
        for (int s : stages) {
            //s는 1과 N+1 사이
            count[s]++;
        }
        
        double total = stages.length;
        HashMap<Integer, Double> fails = new HashMap<>();
        
        for (int i = 1; i <= N; i++) {
            if (count[i] ==0) {
                fails.put(i,0.);
            } else {
                fails.put(i, count[i]/total); //실패율
                total -= count[i]; //분모 빼주기
            }
            
        }
        
        return fails.entrySet().stream().sorted((o1, o2) -> Double.compare(o2.getValue(), o1.getValue())).mapToInt(HashMap.Entry::getKey).toArray();
    }
}